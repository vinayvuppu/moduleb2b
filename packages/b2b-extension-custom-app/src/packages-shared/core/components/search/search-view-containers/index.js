export * from './search-view-containers';
