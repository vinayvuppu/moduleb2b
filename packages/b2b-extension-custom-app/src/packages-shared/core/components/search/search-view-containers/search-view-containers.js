import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import * as searchActions from '../../../actions/search';
import {
  selectIsSearchSliceInitialized,
  selectSearchFilters,
  selectSearchResults,
  selectSearchTotal,
  selectSearchCount,
} from '../../../reducers/search';
import SearchView from '../search-view';

// This component is exported only for testing. Use `SearchViewWithReduxState`.
export class SearchViewContainer extends React.PureComponent {
  static displayName = 'SearchViewContainer';

  static propTypes = {
    initialSliceState: PropTypes.object,
    isSearchSliceInitialized: PropTypes.bool.isRequired,
    searchSliceName: PropTypes.string.isRequired,
    // Action creators
    initializeSearchSlice: PropTypes.func.isRequired,
    setSearchFiltersState: PropTypes.func.isRequired,
  };

  static defaultProps = { initialSliceState: {} };

  componentDidMount() {
    if (!this.props.isSearchSliceInitialized)
      this.props.initializeSearchSlice(
        this.props.initialSliceState,
        this.props.searchSliceName
      );
  }

  handleSetSearchFilterState = opts => {
    this.props.setSearchFiltersState(opts, this.props.searchSliceName);
  };

  render() {
    return this.props.isSearchSliceInitialized ? (
      <SearchView
        {...this.props}
        setSearchFiltersState={this.handleSetSearchFilterState}
      />
    ) : null;
  }
}

const mapStateToProps = (state, ownProps) => {
  const searchSliceName = ownProps.searchSliceName;
  const initialSliceState = ownProps.initialSliceState;
  const isSearchSliceInitialized = selectIsSearchSliceInitialized(
    state,
    searchSliceName
  );
  return isSearchSliceInitialized
    ? {
        isSearchSliceInitialized,
        ...selectSearchFilters(state, searchSliceName),
        results: selectSearchResults(state, searchSliceName),
        total: selectSearchTotal(state, searchSliceName),
        count: selectSearchCount(state, searchSliceName),
      }
    : { isSearchSliceInitialized, initialSliceState };
};
const mapDispatchToProps = {
  initializeSearchSlice: searchActions.initializeSearchSlice,
  setSearchFiltersState: searchActions.setSearchFiltersState,
};

// eslint-disable-next-line import/prefer-default-export
export const SearchViewWithReduxState = connect(
  mapStateToProps,
  mapDispatchToProps,
  undefined,
  // If the parent component manages some state about the table outside of it,
  // and the state changes, the table will not be re-rendered because the
  // (connected) props didn't change, thus `connect` will prevent the child to
  // render again (and therefore the table).
  // For now, until we find a better solution, we disable the equality checks
  // done by `connect`, causing always a re-render.
  { pure: false }
)(SearchViewContainer);
