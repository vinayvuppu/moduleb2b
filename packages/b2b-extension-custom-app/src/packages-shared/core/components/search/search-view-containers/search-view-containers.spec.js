import React from 'react';
import { shallow } from 'enzyme';
import { SearchViewContainer } from './search-view-containers';

const createTestProps = props => ({
  isSearchSliceInitialized: false,
  searchSliceName: 'foo',
  initializeSearchSlice: jest.fn(),
  setSearchFiltersState: jest.fn(),

  // proxied props
  onSearch: jest.fn(),
  filterDefinitions: {
    createdAt: {
      label: 'Created',
      filterTypes: {
        range: {
          filterComponent: jest.fn(),
          tagComponent: jest.fn(),
          label: 'range',
        },
      },
    },
  },
  noResultsText: undefined,
  searchText: null,
  sorting: {
    key: 'createdAt',
    order: 'desc',
  },
  filters: {},
  page: 1,
  perPage: 20,
  results: [],
  total: 0,
  count: 0,

  children: jest.fn(),

  ...props,
});

describe('rendering', () => {
  describe('when initialized', () => {
    let props;
    let wrapper;

    describe('when proxying props', () => {
      beforeEach(() => {
        props = createTestProps({
          isSearchSliceInitialized: true,
          one: 'one',
          two: 'two',
          three: 'three',
        });
        wrapper = shallow(<SearchViewContainer {...props} />);
      });

      it('should output correct tree', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
  describe('when not initialized', () => {
    let props;
    let wrapper;

    beforeEach(() => {
      props = createTestProps({ isSearchSliceInitialized: false });
      wrapper = shallow(<SearchViewContainer {...props} />);
    });

    it('should not render anything', () => {
      expect(wrapper.type()).toEqual(null);
    });
  });
});

describe('interactions', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps({ isSearchSliceInitialized: false });
    wrapper = shallow(<SearchViewContainer {...props} />);
  });

  describe('when search filters change', () => {
    describe('with search text', () => {
      beforeEach(() => {
        wrapper.instance().handleSetSearchFilterState({
          searchText: 'foo',
          sorting: { key: 'createdAt', order: 'asc' },
        });
      });

      it('should invoke `setSearchFiltersState` with sorting key', () => {
        expect(props.setSearchFiltersState).toHaveBeenLastCalledWith(
          {
            searchText: expect.any(String),
            sorting: { key: 'createdAt', order: expect.any(String) },
          },
          expect.any(String)
        );
      });

      it('should invoke `setSearchFiltersState` with sorting order', () => {
        expect(props.setSearchFiltersState).toHaveBeenLastCalledWith(
          {
            searchText: expect.any(String),
            sorting: { key: expect.any(String), order: 'asc' },
          },
          expect.any(String)
        );
      });

      it('should invoke `setSearchFiltersState` with search text', () => {
        expect(props.setSearchFiltersState).toHaveBeenLastCalledWith(
          {
            searchText: 'foo',
            sorting: expect.any(Object),
          },
          expect.any(String)
        );
      });
    });
  });
});

describe('callbacks', () => {
  let props;
  let wrapper;

  describe('when `setSearchFiltersState` is invoked', () => {
    beforeEach(() => {
      props = createTestProps({ isSearchSliceInitialized: true });
      wrapper = shallow(<SearchViewContainer {...props} />);

      wrapper.prop('setSearchFiltersState')({ foo: 'bar' });
    });

    it('should call action creator with the slice name', () => {
      expect(props.setSearchFiltersState).toHaveBeenCalledTimes(1);
      expect(props.setSearchFiltersState).toHaveBeenLastCalledWith(
        expect.any(Object),
        'foo'
      );
    });

    it('should call action creator with the search options', () => {
      expect(props.setSearchFiltersState).toHaveBeenCalledTimes(1);
      expect(props.setSearchFiltersState).toHaveBeenLastCalledWith(
        { foo: 'bar' },
        expect.any(String)
      );
    });
  });
});

describe('mounting', () => {
  let props;
  let wrapper;

  describe('when search slice is initialized', () => {
    beforeEach(() => {
      props = createTestProps({ isSearchSliceInitialized: true });
      wrapper = shallow(<SearchViewContainer {...props} />);
      wrapper.instance().componentDidMount();
    });

    it('should not dispatch `initializeSearchSlice`', () => {
      expect(props.initializeSearchSlice).not.toHaveBeenCalled();
    });
  });
  describe('when search slice is not initialized', () => {
    beforeEach(() => {
      props = createTestProps({ isSearchSliceInitialized: false });
      wrapper = shallow(<SearchViewContainer {...props} />);
      wrapper.instance().componentDidMount();
    });

    describe('without initial state', () => {
      it('should dispatch `initializeSearchSlice`', () => {
        expect(props.initializeSearchSlice).toHaveBeenCalledTimes(1);
        expect(props.initializeSearchSlice).toHaveBeenLastCalledWith({}, 'foo');
      });
    });

    describe('with initial state', () => {
      const initialSliceState = { initial: 'state' };

      beforeEach(() => {
        props = createTestProps({
          isSearchSliceInitialized: false,
          initialSliceState,
        });
        wrapper = shallow(<SearchViewContainer {...props} />);
        wrapper.instance().componentDidMount();
      });

      it('should dispatch `initializeSearchSlice` passing the state', () => {
        expect(props.initializeSearchSlice).toHaveBeenCalledTimes(1);
        expect(props.initializeSearchSlice).toHaveBeenLastCalledWith(
          initialSliceState,
          'foo'
        );
      });
    });
  });
});
