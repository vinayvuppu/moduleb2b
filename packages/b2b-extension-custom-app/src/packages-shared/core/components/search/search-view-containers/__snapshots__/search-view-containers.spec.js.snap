// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering when initialized when proxying props should output correct tree 1`] = `
<injectTracking(withPendingRequests(SearchView))
  count={0}
  filterDefinitions={
    Object {
      "createdAt": Object {
        "filterTypes": Object {
          "range": Object {
            "filterComponent": [MockFunction],
            "label": "range",
            "tagComponent": [MockFunction],
          },
        },
        "label": "Created",
      },
    }
  }
  filters={Object {}}
  initialSliceState={Object {}}
  initializeSearchSlice={[MockFunction]}
  isSearchSliceInitialized={true}
  onSearch={[MockFunction]}
  one="one"
  page={1}
  perPage={20}
  results={Array []}
  searchSliceName="foo"
  searchText={null}
  setSearchFiltersState={[Function]}
  sorting={
    Object {
      "key": "createdAt",
      "order": "desc",
    }
  }
  three="three"
  total={0}
  two="two"
>
  <Component />
</injectTracking(withPendingRequests(SearchView))>
`;
